### PR TITLE
Add implementation for executebuiltin(RunPlugin()) and JSON-RPC Addons.GetAddons

### PR DESCRIPTION
**Pull request by michaelarnauts**
### Functional description
Adds an implementation for:
* `xbmc.executebuiltin`:
  * `RunPlugin()`
  * `PlayMedia()`
  * `PlayerControl()`
* `executeJSONRPC`:
  * `Addons.GetAddons`
* Tests for those implementations

Also:
* Added dummy addon (`plugin.video.example`) that can be used during testing.
* Made a missing `KODI_STUB_RPC_RESPONSES` into a warning instead of an `Exception`. When the directory was there, but there were no files in it, we returned an OK anyway, so it was already optional.

### Reasoning
I needed these missing implementations for running tests in https://github.com/add-ons/service.iptv.manager

### CODE REVIEW

File: xbmc.py [hunk 1]
```diff
@@ -690,26 +690,25 @@ def executeJSONRPC(jsonrpccommand):  # NOSONAR
...
-                return json.dumps(next(stub.get('response')
-                                       for stub in stub_content
-                                       if json_data.get('params')
-                                       and stub.get('request', {}).get('method') == json_data.get('method')
-                                       and stub.get('request', {}).get('params') == json_data.get('params')))
-            except StopIteration:
-                pass
-
-    return '{"id":1,"jsonrpc":"2.0","result":"OK"}'
+    if json_responses:
```
Comment (basrieter) [inline xbmc.py:693]: Why did you change this part? If the call to the API is not handled by `JsonRpcApi().handle(json_data)` and no predefined `KODI_STUB_RPC_RESPONSES` path is given then it should just fail right? Or is there a scenario that would make the old behaviour not work?

File: xbmc.py [hunk 2]
```diff
@@ -746,9 +745,23 @@ def executebuiltin(function):
...
+        command, params = re.search(r'([A-Za-z]+)\(([^\)]+)\)', function).groups()
+    except AttributeError:
+        raise ValueError('Could not process %s', function)
+
+    if command.lower() in ['runplugin', 'playmedia']:
+        addon, route = re.search(r'plugin://([^/]+)(.*)', params).groups()
+        addon_info = addoninfo.read_addon_xml(os.path.join(__add_on_info.kodi_home_path, 'addons', addon, 'addon.xml'))
+        pluginsource = addon_info.get('pluginsource')
+
+        subprocess.Popen(['python', os.path.join(__add_on_info.kodi_home_path, 'addons', addon, pluginsource), route])
```
Comment (basrieter) [inline xbmc.py]: Is there any reason why we should actually execute the Python script? Would it not suffice to just check the path exists and see if it is a valid Python file? I used `py_compile.compile()` in some other script to do that.

File: tests/test_xbmc.py [hunk 3]
```diff
@@ -28,3 +30,20 @@ def test_keyboard_env_var(self):
...
+            id=1
+        )
+        result = json.loads(xbmc.executeJSONRPC(json.dumps(cmd)))
+        self.assertEqual(result.get('jsonrpc'), '2.0')
+        self.assertIsInstance(result.get('result'), dict)
+        self.assertIsInstance(result.get('result').get('addons'), list)
+
+        # Loop over the installed add-ons and execute them
+        for addon in result.get('result').get('addons'):
+            xbmc.executebuiltin('RunPlugin(plugin://%s/?abc=123)' % addon.get('addonid'))
```
Comment (basrieter) [inline tests/test_xbmc.py]: I am wondering whether this actually tests the functionality of the `xbmc.executebuiltin`. I just does the call and nothing else.

File: .env [hunk 4]
```diff
@@ -0,0 +1,2 @@
+KODI_HOME=tests/home
```
Comment (basrieter) [inline .env]: Please don't include this file. I understand you use it for local development, but it should not be in the sources.

**Review by basrieter — CHANGES_REQUESTED** [review]
Thanks for the contribution. Two things:

- the `.env` file should not be submitted
- the `xbmc.executebuiltin` now actually executes a Python script. I don't know if that is preferred? 
  - It might cause a security issue: you don't know what the other script might do if it is running with Sakee
  - Would it not just suffice to check if that file exists and is a valid Python file?

File: xbmc.py (hunk 2 above)
Comment (michaelarnauts) [reply xbmc.py]: I actually need this to be executed for my use case. I'm using this to test IPTV Manager, and it needs to execute other add-ons so they can send back data. I have added two example add-ons in the test-home for this that do this.

I don't think this is a security problem, since the test-home isn't external but added to the sources.

File: tests/test_xbmc.py (hunk 3 above)
Comment (michaelarnauts) [reply tests/test_xbmc.py]: Indeed. This function doesn't return anything in Kodi, so I can't test for this. In my tests, I write a file to a tempdir and check if it exists. I have added this do the tests of sake so it will fail in case an exception is thrown.

File: .env (hunk 4 above)
Comment (michaelarnauts) [reply .env]: You are right. Can I'll add it to .gitignore then.

File: xbmc.py (hunk 1 above)
Comment (michaelarnauts) [reply xbmc.py:693]: The current behaviour is also to just give OK back if there is no file in the rpc_responses path. I'm okay to always error, but this isn't the current behaviour.m and will probably break a lot of current tests since you would need to implement every JSON RPC value.

File: sakee/sakejsonrpc.py [hunk 5]
```diff
@@ -47,6 +47,38 @@ def handle(self, json_data):
...
+
+            :param obj addon_info:   Information about the current Add-on paths.
+            """
+            self._ADDON_INFO = addon_info
+
+        # noinspection PyPep8Naming
+        def GetAddons(self, type=None, content=None, enabled=None, properties=None, limits=None, installed=True):  # NOSONAR
+            """ Gets all available addons. """
+            addons = []
+            for addon in os.listdir(os.path.join(self._ADDON_INFO.kodi_home_path, 'addons')):
```
Comment (michaelarnauts) [inline sakee/sakejsonrpc.py:65]: Note to self: Only process directories.

File: tests/test_xbmc.py (hunk 3 above)
Comment (basrieter) [reply tests/test_xbmc.py]: Ok

File: xbmc.py (hunk 2 above)
Comment (basrieter) [reply xbmc.py]: That is true, but if other add-ons use the `xbmc.executebuiltin(RunPlugin())` to call other add-ons, those add-ons will then also start and use Sakee, and might not designed for that. But perhaps that would be more an issue for those devs that decide to use it.

Comment (basrieter) [reply xbmc.py:763]: True

File: xbmc.py (hunk 1 above)
Comment (basrieter) [reply xbmc.py:693]: True

**Comment by basrieter** [comment]
There is  one more thing that I was thinking:

- the `executebuiltin()` might work the same as the `executeJSONRPC()`: the latter uses your `JsonRpcApi().handle()` which then implements all the possible methods. You created that very nicely. Would it not make sense to make this similar?

**Comment by michaelarnauts** [comment]
> There is one more thing that I was thinking:
> 
> * the `executebuiltin()` might work the same as the `executeJSONRPC()`: the latter uses your `JsonRpcApi().handle()` which then implements all the possible methods. You created that very nicely. Would it not make sense to make this similar?

Sure, makes sense. I can modify it to use an external class for all the implementations.

**Comment by michaelarnauts** [comment]
I actually like how the setup of the Kodi home worked so nice to work with other add-ons. This is one of those things that makes this more then simple xbmc-stubs.

**Review by basrieter — APPROVED** [review]

**Review by basrieter — APPROVED** [review]

**Comment by michaelarnauts** [comment]
I just noticed I also need to mock `PlayMedia`. I'll add this to this branch.

**Review by basrieter — APPROVED** [review]

**Comment by basrieter** [comment]
Thanks for the contribuation: I will push a new PyPi version soon.
